R/combineResults.R

```diff
@@ -1,203 +1,396 @@
+#' and differences in the availability of genes between references may have unpredictable effects.
+#' 
+#' @author Aaron Lun
+#'
+#' @seealso
+#' \code{\link{SingleR}} and \code{\link{classifySingleR}}, for generating predictions to use in \code{results}.
+#'
+#' \code{\link{combineCommonResults}}, for another approach to combining predictions.
+#'
+#' @examples
+#' # Making up data.
+#' ref <- .mockRefData(nreps=8)
+#' ref1 <- ref[,1:2%%2==0]
+#' ref2 <- ref[,1:2%%2==1]
+#' ref2$label <- tolower(ref2$label)
+#'
+#' test <- .mockTestData(ref)
+#'
+#' # Performing classification within each reference.
+#' test <- scater::logNormCounts(test)
+#'
+#' ref1 <- scater::logNormCounts(ref1)
+#' pred1 <- SingleR(test, ref1, labels=ref1$label)
+#'
+#' ref2 <- scater::logNormCounts(ref2)
+#' pred2 <- SingleR(test, ref2, labels=ref2$label)
+#'
+#' # Combining results with recomputation of scores.
+#' combined <- combineRecomputedResults(
+#'     results=list(pred1, pred2), 
+#'     test=test,
+#'     ref=list(ref1, ref2), 
+#'     labels=list(ref1$label, ref2$label))
+#' 
+#' combined[,1:5]
+#'
+#' @export
+#' @importFrom S4Vectors DataFrame selfmatch metadata
+#' @importFrom BiocParallel bpiterate SerialParam
+#' @importFrom BiocNeighbors KmknnParam
+combineRecomputedResults <- function(results, test, ref, labels, quantile=0.8, 
+    assay.type.test="logcounts", assay.type.ref="logcounts", check.missing=TRUE,
+    BNPARAM=KmknnParam(), BPPARAM=SerialParam())
+{
+    all.names <- c(list(colnames(test)), lapply(results, rownames))
+    if (length(unique(all.names)) != 1) {
+        stop("cell/cluster names in 'results' are not identical")
+    }
+    all.nrow <- c(ncol(test), vapply(results, nrow, 0L))
+    if (length(unique(all.nrow)) != 1) {
+        stop("numbers of cells/clusters in 'results' are not identical")
     }
 
-    all.scores <- do.call(cbind, collected.scores)
+    # Finding groups of cells with the same combination of per-reference assigned labels.
+    collated <- lapply(results, function(x) x$labels) 
+    names(collated) <- make.names(seq_along(collated))
+    collated <- DataFrame(collated)
+    groups <- selfmatch(collated)
+    by.group <- split(seq_along(groups), groups)
 
-    output <- DataFrame(scores = I(all.scores), row.names=rownames(results[[1]]))
-    metadata(output)$common.genes <- metadata(results[[1]])$common.genes
+    test <- .to_clean_matrix(test, assay.type=assay.type.test, check.missing=check.missing, msg="test")
+    ref <- lapply(ref, FUN=.to_clean_matrix, assay.type=assay.type.ref, check.missing=check.missing, msg="ref")
 
-    if (has.de) {
-        metadata(output)$de.genes <- do.call(c, collected.de)
-        names(metadata(output)$de.genes) <- de.names
+    available <- c(list(rownames(test)), lapply(ref, rownames))
+    if (length(unique(available)) != 1) {
+        warning("'test' and 'ref' differ in the universe of available genes")
     }
```

Comment:
I'm getting this error thrown.  Am I wrong in thinking we didn't want to force test and reference genes to be made all the same here?  But this forces that the genes of test, ref1, ref2, ref3... are all the same, no?

Reply:
It uses the intersection of refs and test when recalculating scores, so having different genes available for each may creating some funky, unpredictable effects.

Reply:
Just realized it's only a warning and that I indeed had results generated.

Reply:
I can see where that could be an issue, but I also think that there will only very rarely be the exact same genes available.  Maybe a 'message' instead?  And maybe give some version of a percentage of overlap which could inform if this is likely a real issue versus likely not?

Reply:
Well, you subset each so that they only have the common genes between them. Otherwise, the recalculated scores with certain markers not available to given ref(s) are not really comparable.

Reply:
There's a few points here.

The first is that the new combining approach only _recommends_ use of the same universe of genes (before marker detection), while the old approach _required_ use of the same `common.genes` (after marker detection). The former is both gentler and easier to satisfy; IMHO, it's kind of weird to expect to be able to combine annotation results where the relevant genes are not present in all references!

The second is that, yes, we always have the choice between nothing, message, warning or error. Doing nothing is too dangerous (I'll get back to this), while an error is too frustrating for something that might be good enough for a first-pass analysis. So that leaves a message or warning, and then it's just a choice of whether or not you have `Warning:` slapped onto the front or not. I think that's totally fine, it is a warning after all.

Now, why is it dangerous? Imagine that you have one reference where a particular label is defined by high expression of a set of markers that are not present in the other reference. Given that they're not present, they won't be used to recompute the scores, meaning that the recomputed score will be low and the combined assignment will probably fail. A warning is appropriate here, to flag that the user should probably re-run from the top with the intersection, which gives the algorithm a chance to pick markers that *are* actually shared.

At some point, we can add this approach to `trainSingleR` and friends by using a `combine.mode` argument to allow people to determine which mode they want to use. This will give it the same ease of use as the current combining pipeline.

Reply:
I do understand this.  I just don't think that alerting as soon as a single potentially meaningless gene is missing is the best way here, because two reference sets will rarely contain the exactly same genes by default.  Thus users will ALWAYS see the warning (unless they trim like we do inside anyway).  Thus users will eventually ignore this warning.

I think some sort of percentage mismatch would help stress the issue more.  Or even better, a short list of the top markers missing per reference (when possible, so in full SingleR runs or when the user manually provides original trained refs).

Reply:
For example btw, I get the warning when I use the built in HPCA and Blueprint-ENCODE together.  Also when I use the built in ImmGen and MouseRNAseq together.

Reply:
Then the solution must be to trim all references to have the same set of genes. A percentage will not help if the 0.1% of genes lost are the critical ones. I don't want the warnings to try to be too intelligent, it should just be "potential problem" vs "no problem". Are the warnings annoying? Well, they should be, that's why they're there.

Note that the trimming will be done automatically if we incorporate this approach into `trainSingleR` and `classifySingleR`, in which case there will be no warning. But viewed in isolation, emission of a warning from the standalone function seems appropriate.